File: backend/src/datasource/dbdao/configuration.rs
#![allow(dead_code)]
use super::dao::DBDao;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use sqlx::Row;

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct SystemConfigRecord {
    pub id: i64,
    pub key: String,
    pub value: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub created_at: Option<String>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct ProviderConfigRecord {
    pub id: i64,
    pub provider_name: String,
    pub config_key: String,
    pub config_value: Option<String>,
    pub description: Option<String>,
    pub created_at: Option<String>,
}

impl DBDao {
    /// Get system config by ID
    pub async fn get_system_config(&self, id: i64) -> Result<SystemConfigRecord> {
        let row = sqlx::query(
            "SELECT id, key, value, description, category, created_at
             FROM system_config WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let config = SystemConfigRecord {
            id: row.get("id"),
            key: row.get("key"),
            value: row.get("value"),
            description: row.get("description"),
            category: row.get("category"),
            created_at: row.get("created_at"),
        };

        Ok(config)
    }

    /// Search system configs with pagination
    pub async fn search_system_configs(
        &self,
        key: Option<String>,
        category: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<SystemConfigRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, key, value, description, category, created_at
             FROM system_config
             WHERE ($1 IS NULL OR key = $1)
               AND ($2 IS NULL OR category = $2)
             OFFSET $3 LIMIT $4",
        )
        .bind(key)
        .bind(category)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let configs = rows
            .into_iter()
            .map(|row| SystemConfigRecord {
                id: row.get("id"),
                key: row.get("key"),
                value: row.get("value"),
                description: row.get("description"),
                category: row.get("category"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(configs)
    }

    /// Add system config
    pub async fn add_system_config(
        &self,
        key: &str,
        value: &str,
        description: Option<String>,
        category: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO system_config (key, value, description, category)
             VALUES ($1, $2, $3, $4)
             RETURNING id",
        )
        .bind(key)
        .bind(value)
        .bind(description)
        .bind(category)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update system config
    pub async fn update_system_config(
        &self,
        id: i64,
        key: Option<String>,
        value: Option<String>,
        description: Option<String>,
        category: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE system_config
             SET key = COALESCE($1, key),
                 value = COALESCE($2, value),
                 description = COALESCE($3, description),
                 category = COALESCE($4, category)
             WHERE id = $5
             RETURNING id",
        )
        .bind(key)
        .bind(value)
        .bind(description)
        .bind(category)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Get provider config by ID
    pub async fn get_provider_config(&self, id: i64) -> Result<ProviderConfigRecord> {
        let row = sqlx::query(
            "SELECT id, provider_name, config_key, config_value, description, created_at
             FROM provider_config WHERE id = $1",
        )
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        let config = ProviderConfigRecord {
            id: row.get("id"),
            provider_name: row.get("provider_name"),
            config_key: row.get("config_key"),
            config_value: row.get("config_value"),
            description: row.get("description"),
            created_at: row.get("created_at"),
        };

        Ok(config)
    }

    /// Search provider configs with pagination
    pub async fn search_provider_configs(
        &self,
        provider_name: Option<String>,
        config_key: Option<String>,
        page: i64,
        limit: i64,
    ) -> Result<Vec<ProviderConfigRecord>> {
        let offset = (page - 1) * limit;
        let rows = sqlx::query(
            "SELECT id, provider_name, config_key, config_value, description, created_at
             FROM provider_config
             WHERE ($1 IS NULL OR provider_name = $1)
               AND ($2 IS NULL OR config_key = $2)
             OFFSET $3 LIMIT $4",
        )
        .bind(provider_name)
        .bind(config_key)
        .bind(offset)
        .bind(limit)
        .fetch_all(&self.db)
        .await?;

        let configs = rows
            .into_iter()
            .map(|row| ProviderConfigRecord {
                id: row.get("id"),
                provider_name: row.get("provider_name"),
                config_key: row.get("config_key"),
                config_value: row.get("config_value"),
                description: row.get("description"),
                created_at: row.get("created_at"),
            })
            .collect::<Vec<_>>();

        Ok(configs)
    }

    /// Add provider config
    pub async fn add_provider_config(
        &self,
        provider_name: &str,
        config_key: &str,
        config_value: &str,
        description: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "INSERT INTO provider_config (provider_name, config_key, config_value, description)
             VALUES ($1, $2, $3, $4)
             RETURNING id",
        )
        .bind(provider_name)
        .bind(config_key)
        .bind(config_value)
        .bind(description)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }

    /// Update provider config
    pub async fn update_provider_config(
        &self,
        id: i64,
        provider_name: Option<String>,
        config_key: Option<String>,
        config_value: Option<String>,
        description: Option<String>,
    ) -> Result<i64> {
        let row = sqlx::query(
            "UPDATE provider_config
             SET provider_name = COALESCE($1, provider_name),
                 config_key = COALESCE($2, config_key),
                 config_value = COALESCE($3, config_value),
                 description = COALESCE($4, description)
             WHERE id = $5
             RETURNING id",
        )
        .bind(provider_name)
        .bind(config_key)
        .bind(config_value)
        .bind(description)
        .bind(id)
        .fetch_one(&self.db)
        .await?;

        Ok(row.get("id"))
    }
}

File: backend/datasource/dbdao/configuration.go
package dbdao

import (
	"context"
	"errors"
	"time"

	"gorm.io/gorm"
)

type SystemConfigModel struct {
	ID          int64     `gorm:"primarykey;column:id"`
	Key         string    `gorm:"column:key;type:varchar(255);not null;uniqueIndex"`
	Value       string    `gorm:"column:value;type:text"`
	Description string    `gorm:"column:description;type:varchar(500)"`
	Category    string    `gorm:"column:category;type:varchar(100)"`
	CreatedAt   time.Time `gorm:"column:created_at;autoCreateTime:milli"`
	UpdatedAt   time.Time `gorm:"column:updated_at;autoUpdateTime:milli"`
}

func (SystemConfigModel) TableName() string {
	return "system_configs"
}

type SystemConfig struct {
	ID          int64
	Key         string
	Value       string
	Description string
	Category    string
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

func (d *DB) GetSystemConfig(ctx context.Context, id int64) (*SystemConfig, error) {
	var model SystemConfigModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("system config not found")
		}
		return nil, result.Error
	}

	return &SystemConfig{
		ID:          model.ID,
		Key:         model.Key,
		Value:       model.Value,
		Description: model.Description,
		Category:    model.Category,
		CreatedAt:   model.CreatedAt,
		UpdatedAt:   model.UpdatedAt,
	}, nil
}

func (d *DB) SearchSystemConfigs(ctx context.Context, key, category string, page, limit int64) ([]*SystemConfig, int64, error) {
	var models []SystemConfigModel
	q := d.DB().WithContext(ctx)

	if key != "" {
		q = q.Where("key LIKE ?", "%"+key+"%")
	}
	if category != "" {
		q = q.Where("category = ?", category)
	}

	var total int64
	q.Model(&SystemConfigModel{}).Count(&total)

	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}

	cfgs := make([]*SystemConfig, len(models))
	for i, m := range models {
		cfgs[i] = &SystemConfig{
			ID:          m.ID,
			Key:         m.Key,
			Value:       m.Value,
			Description: m.Description,
			Category:    m.Category,
			CreatedAt:   m.CreatedAt,
			UpdatedAt:   m.UpdatedAt,
		}
	}

	return cfgs, total, nil
}

func (d *DB) CreateSystemConfig(ctx context.Context, cfg *SystemConfig) (*SystemConfig, error) {
	model := &SystemConfigModel{
		Key:         cfg.Key,
		Value:       cfg.Value,
		Description: cfg.Description,
		Category:    cfg.Category,
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &SystemConfig{
		ID:          model.ID,
		Key:         model.Key,
		Value:       model.Value,
		Description: model.Description,
		Category:    model.Category,
		CreatedAt:   model.CreatedAt,
		UpdatedAt:   model.UpdatedAt,
	}, nil
}

func (d *DB) UpdateSystemConfig(ctx context.Context, cfg *SystemConfig) (*SystemConfig, error) {
	updates := map[string]interface{}{}

	if cfg.Key != "" {
		updates["key"] = cfg.Key
	}
	if cfg.Value != "" {
		updates["value"] = cfg.Value
	}
	if cfg.Description != "" {
		updates["description"] = cfg.Description
	}
	if cfg.Category != "" {
		updates["category"] = cfg.Category
	}

	result := d.DB().WithContext(ctx).Model(&SystemConfigModel{}).Where("id = ?", cfg.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}

	if result.RowsAffected == 0 {
		return nil, errors.New("system config not found")
	}

	return d.GetSystemConfig(ctx, cfg.ID)
}

type ProviderConfigModel struct {
	ID           int64     `gorm:"primarykey;column:id"`
	ProviderName string    `gorm:"column:provider_name;type:varchar(100);not null;index"`
	ConfigKey    string    `gorm:"column:config_key;type:varchar(255);not null"`
	ConfigValue  string    `gorm:"column:config_value;type:text"`
	Description  string    `gorm:"column:description;type:varchar(500)"`
	CreatedAt    time.Time `gorm:"column:created_at;autoCreateTime:milli"`
	UpdatedAt    time.Time `gorm:"column:updated_at;autoUpdateTime:milli"`
}

func (ProviderConfigModel) TableName() string {
	return "provider_configs"
}

type ProviderConfig struct {
	ID           int64
	ProviderName string
	ConfigKey    string
	ConfigValue  string
	Description  string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

func (d *DB) GetProviderConfig(ctx context.Context, id int64) (*ProviderConfig, error) {
	var model ProviderConfigModel
	result := d.DB().WithContext(ctx).Where("id = ?", id).First(&model)
	if result.Error != nil {
		if result.Error == gorm.ErrRecordNotFound {
			return nil, errors.New("provider config not found")
		}
		return nil, result.Error
	}

	return &ProviderConfig{
		ID:           model.ID,
		ProviderName: model.ProviderName,
		ConfigKey:    model.ConfigKey,
		ConfigValue:  model.ConfigValue,
		Description:  model.Description,
		CreatedAt:    model.CreatedAt,
		UpdatedAt:    model.UpdatedAt,
	}, nil
}

func (d *DB) SearchProviderConfigs(ctx context.Context, providerName, configKey string, page, limit int64) ([]*ProviderConfig, int64, error) {
	var models []ProviderConfigModel
	q := d.DB().WithContext(ctx)

	if providerName != "" {
		q = q.Where("provider_name = ?", providerName)
	}
	if configKey != "" {
		q = q.Where("config_key LIKE ?", "%"+configKey+"%")
	}

	var total int64
	q.Model(&ProviderConfigModel{}).Count(&total)

	result := q.Offset(int((page - 1) * limit)).Limit(int(limit)).Find(&models)
	if result.Error != nil {
		return nil, 0, result.Error
	}

	cfgs := make([]*ProviderConfig, len(models))
	for i, m := range models {
		cfgs[i] = &ProviderConfig{
			ID:           m.ID,
			ProviderName: m.ProviderName,
			ConfigKey:    m.ConfigKey,
			ConfigValue:  m.ConfigValue,
			Description:  m.Description,
			CreatedAt:    m.CreatedAt,
			UpdatedAt:    m.UpdatedAt,
		}
	}

	return cfgs, total, nil
}

func (d *DB) CreateProviderConfig(ctx context.Context, cfg *ProviderConfig) (*ProviderConfig, error) {
	model := &ProviderConfigModel{
		ProviderName: cfg.ProviderName,
		ConfigKey:    cfg.ConfigKey,
		ConfigValue:  cfg.ConfigValue,
		Description:  cfg.Description,
	}

	result := d.DB().WithContext(ctx).Create(model)
	if result.Error != nil {
		return nil, result.Error
	}

	return &ProviderConfig{
		ID:           model.ID,
		ProviderName: model.ProviderName,
		ConfigKey:    model.ConfigKey,
		ConfigValue:  model.ConfigValue,
		Description:  model.Description,
		CreatedAt:    model.CreatedAt,
		UpdatedAt:    model.UpdatedAt,
	}, nil
}

func (d *DB) UpdateProviderConfig(ctx context.Context, cfg *ProviderConfig) (*ProviderConfig, error) {
	updates := map[string]interface{}{}

	if cfg.ProviderName != "" {
		updates["provider_name"] = cfg.ProviderName
	}
	if cfg.ConfigKey != "" {
		updates["config_key"] = cfg.ConfigKey
	}
	if cfg.ConfigValue != "" {
		updates["config_value"] = cfg.ConfigValue
	}
	if cfg.Description != "" {
		updates["description"] = cfg.Description
	}

	result := d.DB().WithContext(ctx).Model(&ProviderConfigModel{}).Where("id = ?", cfg.ID).Updates(updates)
	if result.Error != nil {
		return nil, result.Error
	}

	if result.RowsAffected == 0 {
		return nil, errors.New("provider config not found")
	}

	return d.GetProviderConfig(ctx, cfg.ID)
}
